### PR TITLE
Flatten build tool override resolution (#39)

### LINKED ISSUE #39: Bumpy Road: refactor Tool.execute in src/tools/builder/core/wrapper.rs
Summary
Tool.execute in src/tools/builder/core/wrapper.rs exhibits the “Bumpy Road” antipattern: there are two blocks with nested conditional logic, hitting the project threshold (≥ 2). Deeply nested logic increases cognitive load and change risk.

Location
- File: src/tools/builder/core/wrapper.rs
- Symbol: Tool.execute

Why this matters
- Nested, multi-branch logic entangles responsibilities and reduces readability.
- High indentation depth correlates with defects and makes safe changes harder.
- Flattening control flow improves maintainability and testability without changing behaviour.

Suggested approach (no behaviour change)
- Extract Function: move cohesive responsibilities into small, well‑named helpers. Likely seams:
  - parse and validate input parameters
  - map optional overrides (type/language) into BuildRequirement
  - delegate to builder.analyze/builder.build and shape the output JSON
  - error mapping to ToolError with existing messages
- Replace nested conditionals with guard clauses:
  - Early returns on missing/invalid parameters
  - Small predicate helpers for multi‑branch checks
- Keep all user‑visible strings and the ToolOutput payload identical.

Acceptance criteria
- Tool.execute has no nested multi‑branch blocks (depth ≤ 1).
- Cyclomatic complexity is reduced; function length remains reasonable.
- All existing tests pass; no changes to emitted JSON fields or error messages.
- No functional drift: behaviour and output remain byte‑for‑byte compatible.

Backlinks
- PR: https://github.com/leynos/axinite/pull/18
- Requested by: @leynos (PR #18 review thread)

Scope
- This issue tracks a focused internal refactor; no new features in scope.

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch flattens the override handling in [`src/tools/builder/core/wrapper.rs`](https://github.com/leynos/axinite/blob/f926207eb6666981d7a0750a4a06778fd0f9b518/src/tools/builder/core/wrapper.rs) so `BuildSoftwareTool::execute` no longer carries nested multi-branch conditional blocks.

- Extracts software type override parsing into `resolve_software_type`.
- Extracts language override parsing into `resolve_language`.
- Preserves the existing `ToolError::InvalidParameters` variants and user-visible messages for unknown `type` and `language` values.
- Leaves the analysis, override application, build delegation and output JSON flow unchanged.

Closes #39.

## Review walkthrough

- Start with [`src/tools/builder/core/wrapper.rs`](https://github.com/leynos/axinite/blob/f926207eb6666981d7a0750a4a06778fd0f9b518/src/tools/builder/core/wrapper.rs#L13) to review the two private resolver helpers.
- Then review [`BuildSoftwareTool::execute`](https://github.com/leynos/axinite/blob/f926207eb6666981d7a0750a4a06778fd0f9b518/src/tools/builder/core/wrapper.rs#L54) to confirm the override path is now flat and still runs after `builder.analyze` and before `builder.build`.

## Change Type

- [ ] Bug fix
- [ ] New feature
- [x] Refactor
- [ ] Documentation
- [ ] CI/Infrastructure
- [ ] Security
- [ ] Dependencies

## Linked Issue

Closes #39.

## Validation

- [x] `make all`: passed. This covered formatting checks, Clippy, the GitHub WASM tool build, `cargo nextest run --workspace --features test-helpers --profile default` with 3977 tests passed and 7 skipped, and `cargo test --manifest-path tools-src/github/Cargo.toml` with 5 tests passed.
- [x] `git diff --check`: passed.
- [x] Relevant tests pass: covered by `make all`.
- [ ] Manual testing: not run; this is an internal refactor with no runtime behaviour change intended.

## Security Impact

None. The change does not alter permissions, network calls, secrets, file access, tool execution or sandbox policy.

## Database Impact

None. The change does not add migrations, modify schema or affect PostgreSQL or libSQL behaviour.

## Blast Radius

The blast radius is limited to build-tool parameter override parsing in [`src/tools/builder/core/wrapper.rs`](https://github.com/leynos/axinite/blob/f926207eb6666981d7a0750a4a06778fd0f9b518/src/tools/builder/core/wrapper.rs). A regression would affect optional `type` or `language` overrides for the `build_software` tool, but the accepted values, fallback behaviour and error strings are unchanged.

## Rollback Plan

Revert commit `f926207eb6666981d7a0750a4a06778fd0f9b518`.

## Notes

No execplan is associated with this branch; the branch implements the focused issue #39 refactor directly.

---

**Review track**: B (feature/refactor)

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/tools/builder/core/wrapper.rs](https://github.com/leynos/axinite/pull/175/files/f926207eb6666981d7a0750a4a06778fd0f9b518#diff-b4433146c74e52adc98d687e6a13579073e9db91fbc503f42936f1d82c7a74ab)

Comment on lines +29 to +43

~~~rust
    fn resolve_language(
        override_str: Option<&str>,
        fallback: Language,
    ) -> Result<Language, ToolError> {
        match override_str {
            None => Ok(fallback),
            Some("rust") => Ok(Language::Rust),
            Some("python") => Ok(Language::Python),
            Some("typescript") => Ok(Language::TypeScript),
            Some("bash") => Ok(Language::Bash),
            Some(other) => Err(ToolError::InvalidParameters(format!(
                "unknown language: {other}"
            ))),
        }
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/77987/delta?repo-id=1501128&review-id=175&biomarker=Code+Duplication&filename=src%2Ftools%2Fbuilder%2Fcore%2Fwrapper.rs&method=)
The module contains 2 functions with similar structure: BuildSoftwareTool.resolve_language,BuildSoftwareTool.resolve_software_type

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (2 errors, 3 warnings, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Refactoring PR lacks substantive test coverage for new override resolution helpers; zero tests exercise execute method or three private helpers. | Add unit tests covering override resolution with None/Some parameters, valid/invalid values, and error format verification; create integration tests calling BuildSoftwareTool::execute with type and language overrides. |
| Module-Level Documentation | ❌ Error | The file `src/tools/builder/core/wrapper.rs` lacks module-level documentation at the beginning before use statements. | Add a module-level `//!` docstring before use statements explaining the module's purpose and role in the builder pipeline. |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 0.00% which is insufficient. The required threshold is 80.00%. | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| Testing (Unit And Behavioural) | ⚠️ Warning | The refactoring introduced three new private helper methods in wrapper.rs but these lack dedicated unit tests; tests.rs contains 17 test functions for domain types only, with no tests for BuildSoftwareTool, execute method, or the refactored helpers. | Create comprehensive unit tests for the three new helper methods covering None overrides, valid enum variants, invalid value error handling, and case sensitivity; add integration tests for execute method with various parameter combinations. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | The pull request refactors parameter validation logic but provides no snapshot tests validating error message output or JSON serialisation for invalid parameters. | Add snapshot tests using insta to capture JSON output when invalid parameter overrides are supplied (e.g., type=invalid, language=unknown). Validate error message format and serialised output structure for each error case. |
| Unit Architecture | ❓ Inconclusive | Source files referenced in PR summary cannot be located; repository contains no accessible Rust source code to verify architectural separation. | Confirm repository initialisation with correct branch and ensure source files from PR summary are present before re-running assessment. |

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/tools/builder/core/wrapper.rs](https://github.com/leynos/axinite/pull/175/files/ff42edb442c92c33120beafd5ff05df826416bc7#diff-b4433146c74e52adc98d687e6a13579073e9db91fbc503f42936f1d82c7a74ab)

Comment on lines +243 to +252

~~~rust
    fn resolve_software_type_unknown_value_errors() {
        let result =
            BuildSoftwareTool::resolve_software_type(Some("web_service"), SoftwareType::WasmTool);
        match result.unwrap_err() {
            ToolError::InvalidParameters(msg) => {
                assert_eq!(msg, "unknown type: web_service");
            }
            other => panic!("unexpected error: {:?}", other),
        }
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/77987/delta?repo-id=1501128&review-id=175&biomarker=Code+Duplication&filename=src%2Ftools%2Fbuilder%2Fcore%2Fwrapper.rs&method=)
The module contains 2 functions with similar structure: tests.resolve_language_unknown_value_errors,tests.resolve_software_type_unknown_value_errors

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
src/tools/builder/core/wrapper.rs: 

### What lead to degradation?
The module contains 2 functions with similar structure: tests.execute_invalid_language_override_returns_error,tests.execute_invalid_type_override_returns_error

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

File: src/tools/builder/core/wrapper.rs [hunk 1]
```diff
@@ -114,11 +169,317 @@ impl NativeTool for BuildSoftwareTool {
...
+            error: None,
+            started_at: now,
+            completed_at: now,
+            iterations: 1,
+            validation_warnings: vec![],
+            tests_passed: 1,
+            tests_failed: 0,
+            registered: false,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/tools/builder/core/wrapper.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### src/tools/builder/core/wrapper.rs (1)

`120-150`: _⚠️ Potential issue_ | _🟠 Major_ | _🏗️ Heavy lift_
**Add async integration tests for the `execute` method covering override application and error handling.**
The synchronous unit tests for `resolve_software_type` and `resolve_language` helper functions verify correct error messages and fallback preservation in isolation. However, the async `execute` method itself (lines 122–170) lacks integration tests that exercise the full path: JSON parameter extraction, orchestration of overrides, and error propagation through the tool boundary.
Write async tests that:
- Pass JSON with invalid `type` and verify `ToolError::InvalidParameters("unknown type: ...")` is returned
- Pass JSON with invalid `language` and verify `ToolError::InvalidParameters("unknown language: ...")` is returned
- Omit type/language overrides and assert analysed values propagate through to build input
This covers the changed orchestration path and aligns with the guideline to write unit and behavioural tests for new functionality in Rust.

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@src/tools/builder/core/wrapper.rs` around lines 120 - 150, Add async
integration tests for the execute method to exercise JSON extraction, override
application, and error propagation: write async test cases that call
wrapper.execute(...) with serde_json::json! payloads — one with an invalid
"type" string asserting it returns Err(ToolError::InvalidParameters(msg)) where
msg contains "unknown type: ...", another with an invalid "language" asserting
Err(ToolError::InvalidParameters(msg)) with "unknown language: ...", and a third
that omits "type"/"language" to stub/mock builder.analyze() to return a
requirement and assert that the same analyzed requirement values are used (e.g.,
forwarded into builder.build or reflected in the resulting ToolOutput) to
confirm overrides are preserved; use the existing symbols execute,
resolve_software_type, resolve_language, ToolError, and builder.analyze to
locate code and expect async test functions that await execute and assert error
variants and messages precisely.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (3 errors, 2 warnings, 2 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests are substantive but incomplete relative to review comments. Case-sensitivity tests (WasmTool, PYTHON) are missing. Timestamp fixture bug at line 286 remains unfixed. | Add resolve\_software\_type\_is\_case\_sensitive and resolve\_language\_is\_case\_sensitive tests. Fix line 286: timestamp: Utc::now() → timestamp: now. |
| Module-Level Documentation | ❌ Error | src/setup/prompts/render.rs lacks a module-level docstring. Though a submodule, it must carry documentation explaining its purpose, utility and function per the requirement. | Add a module-level //! docstring to src/setup/prompts/render.rs documenting its purpose as a rendering utility submodule for styled prompt output (headers, steps) within the setup prompts system. |
| Unit Architecture | ❌ Error | Unit Architecture violations: execute() depends on wall-clock time without abstraction; test fixtures use non-deterministic Utc::now(); missing review-requested case-sensitivity tests. | Inject clock abstraction; reuse 'now' variable at line 286; add resolve\_software\_type\_is\_case\_sensitive and resolve\_language\_is\_case\_sensitive tests. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Missing case-sensitivity tests for enum parsers (WasmTool/WASM\_TOOL, Rust/PYTHON variants). Test fixture bug: test\_build\_result uses Utc::now() instead of pre-computed 'now' variable for consistency. | Add resolve\_software\_type\_is\_case\_sensitive and resolve\_language\_is\_case\_sensitive tests. Fix test\_build\_result timestamp to use the 'now' variable instead of Utc::now(). |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Missing required case-sensitivity tests; status snapshot lost structure; wasm assertions removed; test timestamp is nondeterministic. | Add case-sensitivity tests to wrapper.rs. Restore status snapshot clarity. Replace wasm assertions. Fix timestamp in test\_build\_result to reuse `now` variable. |
| Out of Scope Changes check | ❓ Inconclusive | Test fixture changes in snapshot tests and tracing-test removal are tangential but closely related to validation of the refactor; however, the pull request summary indicates these are independent test file adjustments unrelated to issue `#39`'s core refactoring objective. | Clarify whether test snapshot updates and tracing\_test removal in unrelated modules were deliberate changes or inadvertent inclusions; if out-of-scope, isolate them into a separate housekeeping PR. |
| Testing (Property / Proof) | ❓ Inconclusive | No result was produced after verification. Marking as INCONCLUSIVE. | Re-run the check or adjust instructions to produce a final result. |

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

### ❌ Failed check

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Execute tests lack snapshot validation of complete ToolOutput JSON structure. Only fragment assertions exist for individual fields; no comprehensive output serialisation snapshot. | Add assert\_snapshot test for execute\_valid\_params\_returns\_success\_output to snapshot the full ToolOutput JSON, verifying refactored code matches original output byte-for-byte. |

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### src/tools/builder/core/wrapper.rs (2)

`403-413`: _🧹 Nitpick_ | _🔵 Trivial_ | _💤 Low value_
**Replace loop-based test cases with rstest parameterisation for better failure isolation.**
Convert `resolve_software_type_all_valid_values` and `resolve_language_all_valid_values` to use `#[rstest]` with individual test cases. This will run each value as a separate test, providing clearer failure messages and better isolation when a specific case fails.

##### ♻️ Example refactor using rstest

```diff
+use rstest::rstest;
+
-#[test]
-fn resolve_software_type_all_valid_values() {
-    for (value, expected) in [
-        ("wasm_tool", SoftwareType::WasmTool),
-        ("cli_binary", SoftwareType::CliBinary),
-        ("library", SoftwareType::Library),
-        ("script", SoftwareType::Script),
-    ] {
-        let result =
-            BuildSoftwareTool::resolve_software_type(Some(value), SoftwareType::Library);
-        assert_eq!(result.expect("expected valid software type"), expected);
-    }
-}
+#[rstest]
+#[case("wasm_tool", SoftwareType::WasmTool)]
+#[case("cli_binary", SoftwareType::CliBinary)]
+#[case("library", SoftwareType::Library)]
+#[case("script", SoftwareType::Script)]
+fn resolve_software_type_all_valid_values(#[case] value: &str, #[case] expected: SoftwareType) {
+    let result = BuildSoftwareTool::resolve_software_type(Some(value), SoftwareType::Library);
+    assert_eq!(result.expect("expected valid software type"), expected);
+}
```

As per coding guidelines: "Replace duplicated tests with `#[rstest(...)]` parameterised cases in Rust."

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the rest with a brief reason, keep changes minimal, and validate.

In `@src/tools/builder/core/wrapper.rs` around lines 401–413 and 435–446, refactor `resolve_software_type_all_valid_values` and `resolve_language_all_valid_values` to use `#[rstest]` parameterisation instead of internal loops. Add `use rstest::rstest;` to the test imports, then replace the for-loop test with `#[rstest]` followed by `#[case(...)]` attributes for each test value/expected pair, and update the function signature to accept `#[case] value: &str, #[case] expected: SoftwareType` (or `Language` respectively). Ensure each case runs as an individual test and that `cargo nextest run` passes.

Also applies to: 437-446

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@src/tools/builder/core/wrapper.rs` around lines 403 - 413, The tests
resolve_software_type_all_valid_values and resolve_language_all_valid_values
currently iterate via for-loops which hides failing cases; refactor them to
parameterised rstest cases: add use rstest::rstest; and replace each loop test
with a #[rstest] function that has multiple #[case(...)] attributes for each
(value, expected) pair, change the test signature to accept #[case] value: &str
and #[case] expected: SoftwareType (or Language) and call
BuildSoftwareTool::resolve_software_type (and the corresponding resolve_language
function) inside the body asserting equality; this will make each pair an
individual test case while preserving existing assertions.

`1-573`: _⚠️ Potential issue_ | _🟠 Major_ | _🏗️ Heavy lift_
**Decompose this file to satisfy the 400-line limit.**
This file is 573 lines, exceeding the mandatory 400-line guideline by 43%. Extract the test module (lines 210–572) into a separate file, such as `tests/builder_core_wrapper.rs` or `src/tools/builder/core/wrapper/tests.rs`, to bring the main module within the limit.
As per coding guidelines: "Files must not exceed 400 lines in length. Large modules must be decomposed."

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the rest with a brief reason, keep changes minimal, and validate.

In `@src/tools/builder/core/wrapper.rs`, the file is 573 lines (43% over the 400-line limit). Extract the entire `#[cfg(test)] mod tests` block (lines 210–572) into a separate test file. Create `src/tools/builder/core/wrapper/tests.rs` (or use the standard `tests/` directory structure if preferred by the project), move all test code (including imports, helper structs FixedClock/FakeSoftwareBuilder, test functions) into that file, add any necessary imports (`use super::*;` or explicit imports from the parent module), and ensure `cargo nextest run` still passes. Update any module structure as needed to expose necessary items for testing.

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@src/tools/builder/core/wrapper.rs` around lines 1 - 573, The file exceeds the
400-line limit; extract the whole test module starting at the #[cfg(test)] mod
tests block into a separate test file to reduce wrapper.rs size. Move the block
(all fixtures, helper types like FixedClock, FakeSoftwareBuilder, and test
functions that reference BuildSoftwareTool, resolve_override,
resolve_software_type, resolve_language, execute, etc.) into e.g.
src/tools/builder/core/wrapper/tests.rs (or tests/builder_core_wrapper.rs), add
the necessary imports (use super::* or explicit imports of BuildSoftwareTool,
ToolError, BuildRequirement, BuildResult, JobContext, SoftwareBuilder, etc.),
and ensure the new test module compiles by exporting any items from wrapper.rs
that the tests need (make public if required) and running the test suite to
verify nothing breaks.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### src/tools/builder/core/wrapper.rs (1)

`183-508`: _🛠️ Refactor suggestion_ | _🟠 Major_ | _🏗️ Heavy lift_
**Split this module to enforce the 400-line limit.**
Move the `#[cfg(test)] mod tests` block into a separate test module file (for example `wrapper_tests.rs`) and keep this production module under 400 lines.
As per coding guidelines: “Files must not exceed 400 lines in length”.

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@src/tools/builder/core/wrapper.rs` around lines 183 - 508, The file exceeds
400 lines because the inline #[cfg(test)] mod tests block is too large; move
that entire tests block into a new file named wrapper_tests.rs and keep the
production wrapper.rs under 400 lines. Create wrapper_tests.rs containing the
exact contents of the #[cfg(test)] mod tests { ... } block (keeping use super::*
and other imports as-is), and in the original file replace the large tests block
with a single #[cfg(test)] mod wrapper_tests; declaration so the test module is
compiled only in test builds; verify any visibility needed by tests (make helper
functions/types pub(crate) if compile errors appear) and run cargo test to
confirm.

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

I would prefer that timing tests be performed deterministically.

There shout be documentation in the developer's guide or testing strategy document making this clear and also setting out clear patterns for re-usability and repeatability of mocks (and the use of off the shelf mock injection patterns such as mockall/mockable where applicable).

> Clock trait adds production complexity (Arc field, trait def, SystemClock impl) enabling one deterministic timing test, without stated reuse strategy or immediate simplifying effect.

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

Use mockable::Clock instead of custom trait. Project's DI guide explicitly recommends mockable for time handling. Add to dev-dependencies and refactor BuildSoftwareTool to use it.
